### PR TITLE
feat: add interval util functions for priority queue

### PR DESCRIPTION
**Pull request by hi-rustin**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/pull/49018

Problem Summary:

### What changed and how does it work?

Added some functions to get the analysis intervals.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by hi-rustin — COMMENTED** [review]
🔢 Self-check

File: pkg/statistics/handle/autoanalyze/priorityqueue/interval.go [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+}
+
+// getLastFailedAnalysisDuration returns the duration since the last failed analysis.
+func getLastFailedAnalysisDuration(
+	sctx sessionctx.Context,
+	schema, tableName string,
+	partitionNames ...string,
+) (time.Duration, error) {
+	query := ""
+	params := make([]interface{}, 0, len(partitionNames)+3)
```
Comment (hawkingrei) [inline pkg/statistics/handle/autoanalyze/priorityqueue/interval.go:108]:
```suggestion
	params := make([]any, 0, len(partitionNames)+3)
```

File: pkg/statistics/handle/autoanalyze/priorityqueue/interval_test.go [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+	"time"
+
+	"github.com/pingcap/tidb/pkg/session"
+	"github.com/pingcap/tidb/pkg/sessionctx"
+	"github.com/pingcap/tidb/pkg/testkit"
+	"github.com/stretchr/testify/require"
+)
+
+func TestGetAverageAnalysisDuration(t *testing.T) {
+	store, _ := testkit.CreateMockStoreAndDomain(t)
```
Comment (hawkingrei) [inline pkg/statistics/handle/autoanalyze/priorityqueue/interval_test.go:28]:
```suggestion
	store := testkit.CreateMockStore(t)
```

File: pkg/statistics/handle/autoanalyze/priorityqueue/interval_test.go [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+	require.Equal(t, time.Duration(3600)*time.Second, avgDuration)
+	// Non-partitioned table.
+	insertFinishedJob(tk, "example_schema1", "example_table1", "")
+	avgDuration, err = getAverageAnalysisDuration(sctx, "example_schema1", "example_table1", "")
+	require.NoError(t, err)
+	require.Equal(t, time.Duration(3600)*time.Second, avgDuration)
+}
+
+func TestGetLastFailedAnalysisDuration(t *testing.T) {
+	store, _ := testkit.CreateMockStoreAndDomain(t)
```
Comment (hawkingrei) [inline pkg/statistics/handle/autoanalyze/priorityqueue/interval_test.go:61]:
```suggestion
	store := testkit.CreateMockStore(t)
```

**Review by hawkingrei — APPROVED** [review]

File: pkg/statistics/handle/autoanalyze/priorityqueue/interval.go [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+`
+
+// getAverageAnalysisDuration returns the average duration of the last 5 successful analyses for each specified partition.
+func getAverageAnalysisDuration(
+	sctx sessionctx.Context,
+	schema, tableName string,
+	partitionNames ...string,
+) (time.Duration, error) {
+	query := ""
+	params := make([]interface{}, 0, len(partitionNames)+3)
```
Comment (hi-rustin) [inline pkg/statistics/handle/autoanalyze/priorityqueue/interval.go:75]:
```suggestion
	params := make([]any, 0, len(partitionNames)+3)
```

**Review by hi-rustin — COMMENTED** [review]
🔢 Self-check

**Review by time-and-fate — COMMENTED** [review]
Looks like we didn't consider what would happen if there are no such rows in the analyze history.

File: pkg/statistics/handle/autoanalyze/priorityqueue/interval.go [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+	"github.com/pingcap/tidb/pkg/statistics/handle/util"
+	"github.com/pingcap/tidb/pkg/types"
+)
+
+const avgDurationQueryForTable = `
+	SELECT AVG(TIMESTAMPDIFF(SECOND, start_time, end_time)) AS avg_duration
+	FROM mysql.analyze_jobs
+	WHERE table_schema = %? AND table_name = %? AND state = 'finished' AND fail_reason IS NULL AND partition_name = ''
+	ORDER BY id DESC
+	LIMIT 5;
```
Comment (time-and-fate) [inline pkg/statistics/handle/autoanalyze/priorityqueue/interval.go:30]: This SQL is a bit questionable with both an `avg` and `ORDER BY + LIMIT`.

**Comment by hi-rustin** [comment]
> Looks like we didn't consider what would happen if there are no such rows in the analyze history.

Then we will return 0 and compare it with 0 by `0 < 0`, then we can analyze it because it doesn't satisfy the condition.

**Comment by time-and-fate** [comment]
> > Looks like we didn't consider what would happen if there are no such rows in the analyze history.
> 
> Then we will return 0 and compare it with 0 by `0 < 0`, then we can analyze it because it doesn't satisfy the condition.

Have you tested this case to make sure?
`AVG` and `MIN` on empty result set would be `NULL` instead of `0`.

And I think we'd better add some comments. For example, it's not obvious that `len(rows) == 0` in `getLastFailedAnalysisDuration()` is an expected case.

File: pkg/statistics/handle/autoanalyze/priorityqueue/interval.go (hunk 5 above)
Comment (hi-rustin) [reply pkg/statistics/handle/autoanalyze/priorityqueue/interval.go:30]: I want to calculate the average analysis time for the most recent 5 records. Could you please provide more information about the issue that is causing you worry?

Comment (time-and-fate) [reply pkg/statistics/handle/autoanalyze/priorityqueue/interval.go:30]: The semantic of a SELECT statement is generally `JOIN -> WHERE -> AGG -> HAVING -> ORDER BY -> LIMIT`. You would need to write it like `avgDurationQueryForPartition` below to get what you want.
You can see this example: https://www.db-fiddle.com/f/3moV1YrnrZchZTaK41fk2h/1

**Comment by hi-rustin** [comment]
> > > Looks like we didn't consider what would happen if there are no such rows in the analyze history.
> > 
> > 
> > Then we will return 0 and compare it with 0 by `0 < 0`, then we can analyze it because it doesn't satisfy the condition.
> 
> Have we tested this case to make sure? `AVG` and `MIN` on empty result set would be `NULL` instead of `0`.
> 
> And I think we'd better add some comments. For example, it's not obvious that `len(rows) == 0` in `getLastFailedAnalysisDuration()` is an expected case.

I added some tests to cover it and also updated the implementation to make it more obvious.

**Review by time-and-fate — APPROVED** [review]

**Comment by hi-rustin** [comment]
@ti-chi-bot Hi man

**Pull request merged**